ipv8/dht/routing.py

```diff
@@ -0,0 +1,208 @@
+            node.bucket = self
+            self.last_changed = time.time()
+            return True
+
+        return False
+
+    def split(self):
+        if len(self.nodes) < self.max_size:
+            return False
+
+        b_0 = Bucket(self.prefix_id + u'0', self.max_size)
+        b_1 = Bucket(self.prefix_id + u'1', self.max_size)
+        for node in self.nodes.itervalues():
+            if b_0.owns(node.id):
+                b_0.add(node)
+            elif b_1.owns(node.id):
+                b_1.add(node)
+            else:
+                self.logger.error('Failed to place node into bucket while splitting')
+        return b_0, b_1
+
+
+class RoutingTable(object):
+    """
+    The RoutingTable is a binary tree that keeps track of Nodes that we have a connection to.
+    """
+
+    def __init__(self, my_node_id):
+        self.my_node_id = my_node_id
+        self.trie = datrie.Trie(u'01')
+        self.trie[u''] = Bucket(u'')
+        self.lock = RLock()
+
+    def get_bucket(self, node_id):
+        node_id_binary = id_to_binary_string(node_id)
+        return self.trie.longest_prefix_value(unicode(node_id_binary), default=None) or self.trie[u'']
+
+    def add(self, node):
+        with self.lock:
+            bucket = self.get_bucket(node.id)
+
+            # Add/update node
+            if not bucket.add(node):
+                # If adding the node failed, split the bucket
+                # Splitting is only allowed if our own node_id falls within this bucket
+                if bucket.owns(self.my_node_id):
+                    bucket_0, bucket_1 = bucket.split()
+                    self.trie[bucket.prefix_id + u'0'] = bucket_0
+                    self.trie[bucket.prefix_id + u'1'] = bucket_1
+                    del self.trie[bucket.prefix_id]
+
+                    # Retry
+                    return self.add(node)
+            else:
+                return bucket.get(node.id)
+
+    def remove_bad_nodes(self):
+        with self.lock:
+            for bucket in self.trie.values():
+                for node_id, node in bucket.nodes.items():
+                    if node.status == NODE_STATUS_BAD:
+                        bucket.nodes.pop(node_id, None)
+
+    def has(self, node):
+        return bool(self.get_bucket(node.id).get(node.id))
+
+    def closest_nodes(self, node_id, max_nodes=8, exclude=None):
+        with self.lock:
+            hash_binary = unicode(id_to_binary_string(node_id))
+            prefix = self.trie.longest_prefix(hash_binary, default=u'')
+
+            nodes = set()
+            for i in reversed(range(len(prefix) + 1)):
+                for suffix in self.trie.suffixes(prefix[:i]):
+                    bucket = self.trie[prefix[:i] + suffix]
+                    nodes |= {node for node in bucket.nodes.itervalues() if node.status != NODE_STATUS_BAD and \
+                                                                            (not exclude or node.id != exclude.id)}
```

Comment:
Please, make it `(excluded_node not None or node.id != excluded_node.id)` for readability.